Ed-Fi-Exchange-OSS/EdFi-Project-Buzz
# [FIF-300] Adding EdFi.Installer.AppCommon to API and UI apps

### OVERVIEW 

This PR is to complete the installation changes for configuring UI and API applications. They will run on the localhost ports 4200 (UI) and 3000 (API) as per our developer environment.

We are not using the EdFi.Installer.AppCommon to install the apps into IIS because we are using NGINX, whereas ODS tools groups is installing dotnet applications into IIS. End Users can add IIS websites with a reverse proxy to our UI and API ports to achieve an IIS website. 

The next steps will be to spike how to configure the reverse proxies for API and UI in IIS websites or webapps. At present we can run our webs on our familiar ports locally.

Configuration.json is now kept locally only. You can copy the example configuration.json file and overwrite with your own values.

This is the `configuration.json` I use:

```json
{
    "installPath": "C:/Ed-Fi/Buzz",
    "toolsPath": "C:/temp/tools",
    "packagesPath": "C:/temp/packages",
    "artifactRepo": "C:/nugetRepo",
    "installDatabase": true,
    "installEtl": true,
    "installUi": true,
    "installApi": true,
    "idProvider": "google",
    "clientSecret": "*******",
    "googleClientId": "********",
    "googleCallback": "https://localhost:3000",
    "adfsClientId": null,
    "adfsTenantId": null,
    "postgresDatabase": {
        "host": "localhost",
        "port": 5432,
        "username": "postgres",
        "password": "pa55w0rd",
        "database": "edfi_buzz",
        "schema":  "buzz"
    },
    "sqlServerDatabase": {
        "host": "127.0.0.1",
        "port": 49412,
        "username": "ods_user",
        "password": "P@ssw0rd",
        "database": "ods-with-amt",
        "UseIntegratedSecurity":false
    },
    "database": {
        "version": "0.1.0"
    },
    "etl" : {
        "version": "0.1.0",
        "datasourceFormat": "amt",
        "odsDataStandard": "ds3",
        "max": 20,
        "idleTimeoutMilliseconds": 5000,
        "connectionTimeoutMilliseconds": 2000,
        "trustServerCertificate": false,
        "enableArithAbort": true,
        "encrypt": false
    },
    "api": {
        "version": "0.1.0",
        "url": "http://localhost:3000",
        "port": 3000,
        "nginxPort": 3000,
        "surveyFilesFolder": "C:/Ed-Fi/Buzz"
    },
    "ui": {
        "version": "0.1.0",
        "url": "http://localhost:4200",
        "port": 4200,
        "nginxPort": 4200
    }
}

```

### TO TEST

- Run `.\test.ps1` in the root to refresh the NuGet packages in the `C:\nugetRepo` and remove any existing packages in `C:\temp\packages`. This also assumes you use those as the `packagesPath` and `artifactRepo` values in your configuration.json file.

Locally, I use a single line command to jump to the appropriate directories to run the uninstaller and then test scripts in succession. I am using Z-Location in the `z` command to abbreviate when I change paths.:

```powershell
z proj ed buzz installer eng win;.\uninstall.ps1;z proj ed buzz;.\test.ps1
```
- Change to .\EdFi.Buzz.Installer\eng\Windows\ and execute install.ps1. installs all apps. You should have three running EdFi Buzz services
- The service statuses should be output at the end of the install. All three services should be running. 
- UI should be available at http://localhost:4200. It is running on Nginx only now. We are not loading into IIS yet.
- API will be available at http://localhost:300/graphl. 
- Again, you can change to the installer directory to run the installer function.
- Adjusting the install(app name) boolean settings will skip app installations.

### TEST THAT NOT INSTALLING ANY APPS THROWS AN ERROR

- Change the configuration.json to skip installation of all of the apps.
- Verify that a error says you need to install one app.

Your final output should look like this, and ETL is not running because of a bug on buzzContactInfo...

```
2020-10-01 12:27:06,645 INFO  - Installing the service with id 'edfi-buzz-ui'
2020-10-01 12:27:07,091 INFO  - Starting the service with id 'edfi-buzz-ui'
End Ed-Fi Buzz UI installation.
Package installation completed for UI.
Checking Ed-Fi Buzz App Services and Webs statuses ...

Name                           Value
----                           -----
Ed-Fi Buzz UI Website          Status code returned: 200
Ed-Fi Buzz Api Service         Running
Ed-Fi Buzz UI Service          StartPending
Ed-Fi Buzz ETL Service         Stopped
Ed-Fi Buzz API Website         Status code returned: 200
```

### Known Issues

- The ETL starts then stops because of a `buzzContactInfo` task error on startup.

```
Error: Error processing 'C:\Ed-Fi\Buzz\Etl\dist/tasks/buzzContactInfo.js': Invalid task 'buzzContactInfo' - expected function, received object.
    at Object.getTasks [as default] (C:\Ed-Fi\Buzz\Etl\dist\node_modules\graphile-worker\dist\getTasks.js:125:31)
    at async main (C:\Ed-Fi\Buzz\Etl\dist\node_modules\graphile-worker\dist\cli.js:98:26)
```

- We are not creating IIS web sites; only installing the NGINX. A reverse proxy is required and will call for specific configuration.
